simulator/mips/mips_register/mips_register.h

```diff
@@ -28,9 +27,17 @@ class MIPSRegister {
 public:
     static constexpr const size_t MAX_REG = MAX_VAL_RegNum;
 
+    struct InvalidRegNum final : std::runtime_error
+    {
+        explicit InvalidRegNum( const uint8 num)
+                : std::runtime_error( std::string( "Invalid MIPS register num ") + std::to_string( num) + '\n')
+        { }
+    };
+
     explicit MIPSRegister( uint8 id) : MIPSRegister( static_cast<RegNum>( id))
     {
-        assert( id < 32u);
+        if( id >= 32u)
```

Comment:
Assertion is the right thing here.
Assertions check the internal consistency of the program, and they are disabled in a Release build mode. If assertion fails, it says that problem appears due to the programmer's error.
Exceptions are visible for outside users. They say that user's input is incorrect, inconsistent etc.